src/main/java/org/spongepowered/common/service/scheduler/SchedulerBase.java

```diff
@@ -0,0 +1,265 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.service.scheduler;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.Optional;
+import com.google.common.collect.Sets;
+import org.spongepowered.api.plugin.PluginContainer;
+import org.spongepowered.api.service.scheduler.SchedulerQuery;
+import org.spongepowered.api.service.scheduler.Task;
+import org.spongepowered.common.Sponge;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+abstract class SchedulerBase implements SchedulerQuery {
+
+    // The simple queue of all pending (and running) ScheduledTasks
+    private final Map<UUID, ScheduledTask> taskMap = new ConcurrentHashMap<UUID, ScheduledTask>();
+    private final ScheduledTask.TaskSynchroncity syncType;
+    private long sequenceNumber = 0L;
+    private final String taskNameFmt;
+
+    protected SchedulerBase(ScheduledTask.TaskSynchroncity type) {
+        this.syncType = type;
+        this.taskNameFmt = "%s-" + (type == ScheduledTask.TaskSynchroncity.SYNCHRONOUS ? "S" : "A") + "-%d";
+    }
+
+    @Override
+    public Optional<Task> getTaskById(UUID id) {
+        return Optional.<Task>fromNullable(this.taskMap.get(checkNotNull(id, "id")));
+    }
+
+    @Override
+    public Set<Task> getTasksByName(String pattern) {
+        Pattern searchPattern = Pattern.compile(checkNotNull(pattern, "pattern"));
+        Set<Task> matchingTasks = this.getScheduledTasks();
+
+        Iterator<Task> it = matchingTasks.iterator();
+        while (it.hasNext()) {
+            Matcher matcher = searchPattern.matcher(it.next().getName());
+            if (!matcher.matches()) {
+                it.remove();
+            }
+        }
+
+        return matchingTasks;
+    }
+
+    @Override
+    public Set<Task> getScheduledTasks() {
```

Comment:
You can make the return type here more specific so that you don't to cast to SchedulerTask elsewhere